### PR TITLE
Improve ArrayRef

### PR DESCRIPTION
**Pull request by smessmer**
Summary: Mostly making some stuff in ArrayRef constexpr to give it better perf.

Differential Revision: D8926785

### CODE REVIEW

**Comment by colesbury** [comment]
In what sort of use cases did you see a performance improvement?

**Comment by smessmer** [comment]
Sorry, I probably formulated that a bit misleadingly. It doesn't improve perf for existing use cases, but it allows ArrayRef to be used in constexpr environments in future (for example for c10).

Constexpr per se doesn't improve perf. Some expressions can be evaluated at compile time and some cannot. For expressions that can be evaluated at compile time, many compilers will do so, even if you don't add the "constexpr" keyword. What the "constexpr" keyword does is it asserts that the expression can be evaluated at compile time. So it makes sure that the stuff you want to be run at compile time actually can be done so.

File: aten/src/ATen/ArrayRef.h [hunk 1]
```diff
@@ -23,170 +24,184 @@
...
+  /// The number of elements.
+  size_type Length;
+
+ public:
+  /// @name Constructors
+  /// @{
+
+  /// Construct an empty ArrayRef.
+  // TODO Make this explicit
+  constexpr ArrayRef() : Data(nullptr), Length(0) {}
```
Comment (ezyang) [inline aten/src/ATen/ArrayRef.h]: C++ noob question: why does explicit matter here? There's no implicit conversion to have created?

Comment (ezyang) [reply aten/src/ATen/ArrayRef.h]: Also, why did you delete the `/* implicit */` comments? They seem like useful reminders.

File: aten/src/ATen/ArrayRef.h [hunk 2]
```diff
@@ -23,170 +24,184 @@
...
+
+  /// Construct an ArrayRef from a single element.
+  constexpr ArrayRef(const T& OneElt) : Data(&OneElt), Length(1) {}
+
+  /// Construct an ArrayRef from a pointer and length.
+  constexpr ArrayRef(const T* data_, size_t length)
```
Comment (ezyang) [inline aten/src/ATen/ArrayRef.h]: The renaming of the argument here seems both gratuitous, and wrong (aren't underscore suffixed identifiers supposed to be members?!)

File: aten/src/ATen/ArrayRef.h [hunk 3]
```diff
@@ -23,170 +24,184 @@
...
+  }
+
+  /// size - Get the array size.
+  constexpr size_t size() const {
+    return Length;
+  }
+
+  /// front - Get the first element.
+  C10_CPP14_CONSTEXPR const T& front() const {
+    AT_CHECK(!empty(), "Empty list!");
```
Comment (ezyang) [inline aten/src/ATen/ArrayRef.h]: The old error message seemed much more informative than the new one

File: aten/src/ATen/ArrayRef.h [hunk 4]
```diff
@@ -23,170 +24,184 @@
...
+
+  /// equals - Check for element-wise equality.
+  constexpr bool equals(ArrayRef RHS) const {
+    return Length == RHS.Length && std::equal(begin(), end(), RHS.begin());
+  }
+
+  /// slice(n, m) - Chop off the first N elements of the array, and keep M
+  /// elements in the array.
+  C10_CPP14_CONSTEXPR ArrayRef<T> slice(size_t N, size_t M) const {
+    AT_CHECK(N + M <= size(), "N = ", N, "; M = ", M, "; size = ", size());
```
Comment (ezyang) [inline aten/src/ATen/ArrayRef.h]: Ditto here; why would you remove perfectly good explanatory text?

**Comment by ezyang** [comment]
Code organization question: in this PR, you moved C++17.h from caffe2 folder to aten folder. But isn't this file supposed to be shared between both Caffe2 and PyTorch?

It's inappropriate to put C++17.h in ATen directly, because ATen is too heavy a dependency for us to put in mobile. It's best for us to have some directory which doesn't include anything outside of it, which is OK to ship to mobile (and we promise to keep an eye on binary sizes here). This directory should be depended on by both ATen and Caffe2, and in the near future Tensor and Storage classes will also live in this directory.

File: aten/src/ATen/ArrayRef.h (hunk 1 above)
Comment (smessmer) [reply aten/src/ATen/ArrayRef.h]: It doesn't. This comment was meant to be 3 lines down at the one-element constructor. I fixed it.

File: aten/src/ATen/ArrayRef.h (hunk 2 above)
Comment (smessmer) [reply aten/src/ATen/ArrayRef.h]: fixed

File: aten/src/ATen/ArrayRef.h (hunk 1 above)
Comment (smessmer) [reply aten/src/ATen/ArrayRef.h]: About the `implicit` comments, explicit/implicit only matters for one-argument constructors. I could leave the `implicit` comment at the `OneElt` constructor, but that's actually the one with the TODO to make explicit, that's why I removed the comment.

File: aten/src/ATen/ArrayRef.h (hunk 3 above)
Comment (smessmer) [reply aten/src/ATen/ArrayRef.h]: fixed

File: aten/src/ATen/ArrayRef.h (hunk 4 above)
Comment (smessmer) [reply aten/src/ATen/ArrayRef.h]: tbh, I have no idea why and how this was changed.

**Comment by smessmer** [comment]
About code organization, caffe2 already has this weird pattern that they include .cpp files on a per-file basis to the mobile build. There is no clear separation in the folder structure for files that are or aren't used in mobile. I'm, however, introducing an ATen/utils folder and plan to put this file there. It's introduced in https://github.com/pytorch/pytorch/pull/9897 and I will move ArrayRef there.

I wasn't doing it in this PR, because I didn't figure out yet how to set the build system up for this.

**Comment by ezyang** [comment]
Update for those paying attention: we're going to call the directory `ATen/core` now.

**Pull request closed without merging**